File: lessons/compiled_lang/go/src/f00.loop.go
// testbox: title="range over slice (value form)"

package main

import (
	"fmt"
	"os"
)

func main() {
	// os.ReadDir() already returns entries sorted by filename
	entries, err := os.ReadDir("dirtest")
	if err != nil {
		panic(err)
	}

	for _, entry := range entries {
		if entry.IsDir() {
			fmt.Println(entry.Name() + " is a directory")
		} else {
			fmt.Println(entry.Name() + " is not a directory")
		}
	}
}

File: lessons/compiled_lang/cpp/src/f00.loop.cpp
// testbox: title="range-based for over std::filesystem::directory_iterator"
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

namespace fs = std::filesystem;

int main() {
    // directory_iterator's order isn't guaranteed by the filesystem, so
    //  collect names into a vector and sort first
    std::vector<std::string> items;
    for (const auto& entry : fs::directory_iterator("dirtest"))
        items.push_back(entry.path().filename().string());
    std::sort(items.begin(), items.end());

    // range-based for loop (C++11)
    for (const auto& item : items) {
        if (fs::is_directory("dirtest/" + item))
            std::cout << item << " is a directory" << std::endl;
        else
            std::cout << item << " is not a directory" << std::endl;
    }

    return 0;
}
